src/jit/hwintrinsicxarch.cpp

```diff
@@ -259,112 +226,218 @@ int Compiler::numArgsOfHWIntrinsic(GenTreeHWIntrinsic* node)
+//     true if the node has an imm operand; otherwise, false
+bool HWIntrinsicInfo::isImmOp(NamedIntrinsic id, const GenTree* op)
 {
-    assert(intrinsic != NI_Illegal);
-    assert(intrinsic > NI_HW_INTRINSIC_START && intrinsic < NI_HW_INTRINSIC_END);
-    assert(type >= TYP_BYTE && type <= TYP_DOUBLE);
-    return hwIntrinsicInfoArray[intrinsic - NI_HW_INTRINSIC_START - 1].ins[type - TYP_BYTE];
+    if (HWIntrinsicInfo::lookupCategory(id) != HW_Category_IMM)
+    {
+        return false;
+    }
+
+    if (!HWIntrinsicInfo::MaybeImm(id))
+    {
+        return true;
+    }
+
+    if (genActualType(op->TypeGet()) != TYP_INT)
+    {
+        return false;
+    }
+
+    return true;
 }
 
 //------------------------------------------------------------------------
-// categoryOfHWIntrinsic: get the category of the given intrinsic
+// lookupImmUpperBound: Gets the upper bound for the imm-value of a given NamedIntrinsic
 //
 // Arguments:
-//    intrinsic -- id of the intrinsic function.
+//    id -- The NamedIntrinsic associated with the HWIntrinsic to lookup
 //
 // Return Value:
-//     the category of the given intrinsic
+//     The upper bound for the imm-value of the intrinsic associated with id
 //
-HWIntrinsicCategory Compiler::categoryOfHWIntrinsic(NamedIntrinsic intrinsic)
+int HWIntrinsicInfo::lookupImmUpperBound(NamedIntrinsic id)
 {
-    assert(intrinsic != NI_Illegal);
-    assert(intrinsic > NI_HW_INTRINSIC_START && intrinsic < NI_HW_INTRINSIC_END);
-    return hwIntrinsicInfoArray[intrinsic - NI_HW_INTRINSIC_START - 1].category;
+    assert(HWIntrinsicInfo::lookupCategory(id) == HW_Category_IMM);
+
+    switch (id)
+    {
+        case NI_AVX_Compare:
+        case NI_AVX_CompareScalar:
+        {
+            assert(!HWIntrinsicInfo::HasFullRangeImm(id));
+            return 31; // enum FloatComparisonMode has 32 values
+        }
+
+        default:
+        {
+            assert(HWIntrinsicInfo::HasFullRangeImm(id));
+            return 255;
+        }
+    }
 }
 
 //------------------------------------------------------------------------
-// HWIntrinsicFlag: get the flags of the given intrinsic
+// isFullyImplementedIsa: Gets a value that indicates whether the InstructionSet is fully implemented
 //
 // Arguments:
-//    intrinsic -- id of the intrinsic function.
+//    isa - The InstructionSet to check
 //
 // Return Value:
-//     the flags of the given intrinsic
+//    true if isa is supported; otherwise, false
+bool HWIntrinsicInfo::isFullyImplementedIsa(InstructionSet isa)
+{
+    switch (isa)
+    {
+        // These ISAs have no implementation
+        case InstructionSet_AES:
+        case InstructionSet_BMI1:
+        case InstructionSet_BMI2:
+        case InstructionSet_PCLMULQDQ:
+        {
+            return false;
+        }
+
+        // These ISAs are partially implemented
+        case InstructionSet_AVX:
+        case InstructionSet_AVX2:
+        case InstructionSet_SSE42:
+        {
+            return true;
```

Comment:
Seems like this should have a `TODO-XArch" comment that, approaching the next release, these need to either be fully implemented, or only return true if `COMPlus_EnableIncompleteISAClass` is set.

Reply:
I would imagine that, come next release, we will filter out the unsupported APIs from the reference assembly again (rather than changing this function).

I'll log a tracking bug in CoreFX.

Reply:
https://github.com/dotnet/corefx/issues/30091